erasure: replace bitrot protection mechanism

aead opened the pull request:
This change:
 - Replaces the hash.Hash based bit-rot protection mechanism with a more generic approach (`bitrot.Hash`) - see `/pkg/bitrot`
  - Adds a implementation for [Poly1305](https://en.wikipedia.org/wiki/Poly1305) - see `vendor/github.com/aead/poly1305`
 - Makes poly1305 the default bit-rot protection algorithm
 - Provide new implementations for the file operations (create, read, heal) of the XL backend
 - Add new tests for those file operations
  
## Description
The change does basically two this:
 - It introduces a new bit-rot protection approach - based on polynomial authenticators. To provide a generic interface for bit-rot protection and support hash functions as well as poly. authenticators the general concept of detection data-manipulation is abstracted by the new `bitrot` package. Further the new concept makes it very easy to support new algorithms / different implementations. 

- The second part are the erasure file operations (create, read, heal). This change provides new implementations for those operations and adds more test cases.

## Motivation and Context
 - This change will reduce the CPU load while protecting/verifying erasure-coded data.
 - Simplify erasure file operation implementations. 

## How Has This Been Tested?
New test cases are added for create, read and heal.
All new / existing tests pass.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Fixes #4649 
Fixes #4359

Commit:
erasure: replace hash-based bitrot detection with generic mechanism

This change creates the new pkg/bitrot package which provides a
generic API for bitrot protection mechanisms. Further this change replaces
the hash-function based bitrot protection approach with the more generic one
from the pkg/bitrot.

The bitrot package introduces the polynomial-based authenticators (poly1305
and GHASH) as an additional package. Both authenticators have some advantages over
general purpose hash functions - e.g. they are much faster.

This change adds one implementation of a polynomial authenticator (poly1305) to minio's
vendor folder. See vendor/github.com/aead/

This change also provides new implementations for creating erasure encoded files,
read from erasure encoded files and heal erasure encoded data.

Commit:
add io.Reader and io.Writer converter for StorageAPI

This change starts adding io.Reader and io.Writer converters for
the StorageAPI interface.
This can simplify logic because certain functions can use the functions from
the golang io std instead of custom implementations.

harshavardhana reviewed (approved):
We reviewed this previously and merged. LGTM :+1:

harshavardhana reviewed (changes requested):
We need to make one change undoing my :+1: we haven't bumped the XL version number in this one. We need to address that and also test backward compatible changes.

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,125 @@
+package cmd
```
Header missing

aead replied:
Done

fwessels commented on pkg/bitrot/bitrot.go:
```diff
@@ -0,0 +1,151 @@
+var names = []string{
+	Poly1305:   "poly1305",
+	GHASH:      "ghash",
+	BLAKE2b512: "blake2b",
```
Consider renaming constant `BLAKE2b512' to 'BLAKE2b' in order to match the string name (it is now the odd one out here, and blake2b is what is current written in the json).

aead replied:
Done

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,135 @@
+// MissingShard indicates a missing data/parity shard for erasure coding
+var MissingShard []byte // zero value is nil
+
+// ErasureFileInfo contains information about a erasure file operation (create, read, heal).
```
about a erasure --> about an erasure

aead replied:
Done

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,135 @@
+}
+
+// XLStorage represents an array of disks.
+// The disks contains erasure coded and bitrot-protected data.
```
contains --> contain

aead replied:
Done

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,135 @@
+	return s0
+}
+
+// AppendWriters creates one writer which for every disk of the storage array.
```
remove which

aead replied:
Done

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,135 @@
+}
+
+// AppendWriters creates one writer which for every disk of the storage array.
+// Every writer append data to the corresponding storage disk.
```
append --> appends

aead replied:
Done

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,135 @@
+// ErasureEncode encodes the given data and returns the erasure-coded data.
+// It returns an error if the erasure coding failed.
+func (s XLStorage) ErasureEncode(data []byte) ([][]byte, error) {
+	encoder, err := reedsolomon.New(len(s)/2, len(s)/2)
```
We should be using xlObjects.dataBlocks and xlObjects.parityBlocks here.

aead replied:
Done

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,135 @@
+// ErasureDecode decodes the given erasure-coded data.
+// It returns an error if the decoding failed.
+func (s XLStorage) ErasureDecode(data [][]byte) error {
+	decoder, err := reedsolomon.New(len(s)/2, len(s)/2)
```
Idem as above.

aead replied:
Done

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,135 @@
+		return Errorf("failed to create erasure coding: %v", err)
+	}
+	if err = decoder.Reconstruct(data); err != nil {
+		return Errorf("failed to split data: %v", err)
```
failed to reconstruct data

aead replied:
Done

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,151 @@
+	verified  bool
+}
+
+// Verify returns true iff the computed checksum of the verifier matches the the checksum provided when the verifier
```
typo in iff

aead replied:
`iff` is the short form for `if and only if`... (a common crypto doc abbreviation)  - I can write it out if this is not clear...

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,151 @@
+	verified  bool
+}
+
+// Verify returns true iff the computed checksum of the verifier matches the the checksum provided when the verifier
```
the repeated twice

aead replied:
same as above

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,151 @@
+	return subtle.ConstantTimeCompare(v.Sum(nil), v.sum) == 1
+}
+
+// IsVerified returns true iff Verify was called at least one.
```
iff....
one --> once

aead replied:
same..
Done

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+// IsVerified returns true iff Verify was called at least one.
+func (v *BitrotVerifier) IsVerified() bool { return v.verified }
```
Wouldn't you want to call Verify() implicitly as part of IsVerified?

aead replied:
Done

fwessels commented on cmd/erasure-createfile.go:
```diff
@@ -17,128 +17,76 @@
-			} // else we have reached EOF after few reads, no need to
-			// add an additional 0bytes at the end.
-			break
+// CreateFile creates a new bitrot encoded file spreed over all available disks. CreateFile will create
```
spreed --> spread

aead replied:
Done

fwessels commented on cmd/erasure-createfile.go:
```diff
@@ -17,128 +17,76 @@
-			break
+// CreateFile creates a new bitrot encoded file spreed over all available disks. CreateFile will create
+// the file at the given volume and path. It will read from src until an io.EOF occurs. The given algorithm will
+// be used to protected the erasure encoded file. The random reader should return random data (if it's nil the system PRNG will be used).
```
to protected --> to protect

aead replied:
Done

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,167 @@
+}
+
+// ErasureDecodeDataBlocks decodes the given erasure-coded data.
+// It only decodes the data and parity blocks but does not verify them.
```
I'd remove "and parity" as that is what we want...

aead replied:
Okay, sure - done.

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,167 @@
+	}
+	if ok, err := s.erasure.Verify(data); !ok || err != nil {
+		if err != nil {
+			return Errorf("failed to verify data: %v", err)
```
why not use else if !ok { } further down below are remove the double comparison for err != nil ?

aead replied:
Because there are two different error conditions:
1. for example if we pass two few shards -> ErrTooFewShards
2. the data verification failed because e.g. the data is (after decode) is not correct.
If we remove one check we return the "wrong" error...

We need no !ok because if we enter the first if (`!ok || err != nil`)  but not the second (`err != nil`)
this implies that `ok == false`. 
Or didn't I get your point...

fwessels commented on cmd/erasure-readfile.go:
```diff
@@ -17,328 +17,142 @@
-	rs, err := reedsolomon.New(dataBlocks, parityBlocks)
-	if err != nil {
-		return traceError(err)
+	if err = reduceReadQuorumErrs(errs, []error{}, (s.dataBlocks+s.parityBlocks)/2); err != nil {
```
Use xlObjects.readQuorum for read quorum

aead replied:
Done

fwessels commented on cmd/erasure-readfile.go:
```diff
@@ -17,328 +17,142 @@
+		if length < writeLength {
+			writeLength = length
 		}
+		n, err := writeDataBlocks(writer, blocks, (s.dataBlocks+s.parityBlocks)/2, startOffset, writeLength)
```
We should be using implicit read quorum of `(s.dataBlocks+s.parityBlocks)/2` here
(applies multiple times in this function)

aead replied:
in this case it must be just DataBlocks

But we are computing the chunkisze here, so we must use the avg of dataBlocks and parityBlocks

aead replied:
Never mind - using dataBlocks is the obvious fix for erasure coding if data != parity - done.

fwessels commented on cmd/erasure-readfile.go:
```diff
@@ -17,328 +17,142 @@
-	ok, err := rs.Verify(enBlocks)
-	if err != nil {
-		return traceError(err)
+func erasureReadBlocksConcurrent(disks []StorageAPI, volume, path string, offset int64, blocks [][]byte, verifiers []*BitrotVerifier, errors []error, locks []chan error) {
```
Please add comments (even though private function)

aead replied:
done.

fwessels commented on cmd/erasure-readfile.go:
```diff
@@ -17,328 +17,142 @@
 
-	// Success.
-	return nil
+func erasureReadFromFile(disk StorageAPI, volume, path string, offset int64, buffer []byte, verifier *BitrotVerifier, lock chan<- error) {
```
Please add comments (even though private function)

aead replied:
done.

fwessels commented on cmd/erasure-readfile.go:
```diff
@@ -17,328 +17,142 @@
+		errors[i] = <-locks[i]
+		if errors[i] != nil {
+			disks[i] = OfflineDisk
+			blocks[i] = MissingShard
```
Is this thread safe? (assigning to blocks[i])

aead replied:
Added a comment - done.

Commit:
posix: improve read and append/write to file

This change removes the buffer pool request in AppendFile - it's not necessary
and causes significant memory allocations.
Further this change eliminates on syscall (Seek). This leads to significant performance
improvements.

Both code paths are bottlenecks - detected while benchmarking the erasure (XL) backend with pprof.

Commit:
erasure: update vendored erasure coding reedsolomon

Update erasure coding package klauspost/reedsolomon to the last version.
This will improve create file performance because reedsolomon.Split won't alloc anymore.

Commit:
bitrot: prevent poly1305 from using zero keys

This change prevents poly1305 from using a zero r part for the key.
In this case the poly1305 multiply would have no effect. So we map the
key to 0xFFFFFF... if it is zero.

harshavardhana commented on cmd/xl-v1-metadata.go:
```diff
@@ -50,7 +51,17 @@ func init() {
+		// If the r part of the key (key[:16]) is zero, the poly1305 multiply has no effect. It's
+		// very unlikely because the key will be randomly generated but we should prevent this.
+		// In this case me map the zero key to the inverse one (11111...).
+		if bytes.Equal(key[:16], pKey[:16]) {
```
Quite a tricky code to understand the gist would require little more explanation IMHO.

aead replied:
Okay maybe something like this?

>>Poly1305 uses the first 16 byte of the key for multiplication. If this part of the key is ever zero the multiply has no effect on the checksum. It is very unlikely that this will ever happen because the key is randomly generated, but even tough we should prevent this from happening. If the part of the key is zero, we now use the inverse - all bit's set. This will guarantee, that the multiplication effects the checksum.

harshavardhana replied:
Yes that would be better.. @aead

fwessels commented on cmd/storage-interface.go:
```diff
@@ -48,3 +51,48 @@ type StorageAPI interface {
 	ReadAll(volume string, path string) (buf []byte, err error)
 }
+
+// storageReader is a io.Reader view of a disk
```
a -> an

aead replied:
Done

fwessels commented on cmd/storage-interface.go:
```diff
@@ -48,3 +51,48 @@ type StorageAPI interface {
+		err = io.EOF
+	}
+	return int(n), err
+
```
empty line

aead replied:
Done

fwessels commented on cmd/storage-interface.go:
```diff
@@ -48,3 +51,48 @@ type StorageAPI interface {
+	if err == io.ErrUnexpectedEOF && nn > 0 {
+		err = io.EOF
+	}
+	return int(n), err
```
n is already an int ? 
Or you can even `return` ?

aead replied:
Done

fwessels commented on cmd/storage-rpc-server-datatypes.go:
```diff
@@ -79,11 +81,13 @@ type ReadFileWithVerifyArgs struct {
-	Algo HashAlgo
+	Algo bitrot.Algorithm
+
+	// Stored key used to compute the hash value.
```
Add this: (in case needed by algorithm)

aead replied:
Done

fwessels commented on cmd/erasure.go:
```diff
@@ -0,0 +1,167 @@
+// The disks contain erasure coded and bitrot-protected data.
+type XLStorage struct {
+	disks                    []StorageAPI
+	erasure                  reedsolomon.Encoder
```
Consider moving `erasure` to xlObjects and removing XLStorage altogether (too much overlap)

aead replied:
Discussed - Decided to keep it separated because of some corner case problems - revisit in future...

Commit:
erasure: use a single key per object for bitrot protection

This change moves the key field for bitrot protection from one checksum info
into a new struct (BitrotInfo).

fwessels commented on cmd/xl-v1-metadata.go:
```diff
@@ -17,20 +17,61 @@
 )
 
+// DefaultBitrotAlgorithm is the default algorithm used for bitrot protection
+const DefaultBitrotAlgorithm = bitrot.Poly1305
```
We need to test on ARM whether Poly1305 is faster than SHA256 (which is accelerated)

aead replied:
In general yes - on arm64 github.com/aead/poly1305 will use the reference implementation. If this is slower than SHA-256, we can implement it in assembly: See [sum_arm.s](https://github.com/golang/crypto/blob/master/poly1305/sum_arm.s)

fwessels replied:
The reference implementation will **not** be fast for ARM since the go compiler has some way to go for ARM.
And especially also because the SHA256 uses the crypto instructions to speed up hashing speed to BLAKE2 like performance.
We need to get some benchmark numbers in order to make a decision

Commit:
erasure: fix minor code-style issues

fwessels commented on cmd/erasure-readfile.go:
```diff
@@ -17,328 +17,151 @@
-	// Success.
-	return nil
+// erasureReadReadFromFile reads data from the disk to buffer in parallel.
+// It sends the returned error through the lock channel.
```
ReadRead repeated twice.
